core/src/main/scala/kafka/admin/AclCommand.scala

```diff
@@ -0,0 +1,241 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package kafka.admin
+
+import java.io.{File, FileInputStream}
+import java.util.Properties
+
+import joptsimple._
+import kafka.security.auth._
+import kafka.server.KafkaConfig
+import kafka.utils._
+import org.apache.kafka.common.security.auth.KafkaPrincipal
+import org.apache.kafka.common.utils.Utils
+
+object AclCommand {
+
+  val delimeter = ","
+  val nl = scala.util.Properties.lineSeparator
+
+  def main(args: Array[String]): Unit = {
+
+    val opts = new AclCommandOptions(args)
+
+    opts.checkArgs()
+
+    val actions = Seq(opts.addOpt, opts.removeOpt, opts.listOpt).count(opts.options.has _)
+    if (actions != 1)
+      CommandLineUtils.printUsageAndDie(opts.parser, "Command must include exactly one action: --list, --add, --remove.")
+
+    val props = new Properties()
+    props.load(new FileInputStream(new File(opts.options.valueOf(opts.config))))
+
+    if (!props.containsKey(KafkaConfig.AuthorizerClassNameProp))
+      CommandLineUtils.printUsageAndDie(opts.parser, "Config file must set property = " + KafkaConfig.AuthorizerClassNameProp
+        + " to a class that implements Authorizer interface.")
+
+    val kafkaConfig = KafkaConfig.fromProps(props)
+    val authZ: Authorizer = CoreUtils.createObject(kafkaConfig.authorizerClassName)
+    authZ.configure(kafkaConfig.originals())
+
+    try {
+      if (opts.options.has(opts.addOpt))
+        addAcl(authZ, opts)
+      else if (opts.options.has(opts.removeOpt))
+        removeAcl(authZ, opts)
+      else if (opts.options.has(opts.listOpt))
+        listAcl(authZ, opts)
+    } catch {
+      case e: Throwable =>
+        println("Error while executing topic command " + e.getMessage)
+        println(Utils.stackTrace(e))
+    }
+  }
+
+  private def addAcl(authZ: Authorizer, opts: AclCommandOptions): Unit = {
+    val acls: Set[Acl] = getAcl(opts)
+
+    if (acls.isEmpty)
+      CommandLineUtils.printUsageAndDie(opts.parser, "You must specify one of : --allowprincipals, --denyPrincipals when trying to add acls.")
```

Comment:
It's fine not to default principal to ALL when host is specified. It just that in the current code, if only host is specified, then the principal list is empty. Since the acls are generated from the combination of hosts and principals, the acl list will be empty.